File: javascript/4-OnLogN_merge_sort.js
/* 
    OnLogN_merge_sort.js
    Big-O: O(n log n)
    Classic merge-sort: log n levels of splitting × n work each level. 
*/

import { performance } from "node:perf_hooks";

const mergeSort = (arr) => {
  if (arr.length <= 1) return arr;
  const mid = arr.length >> 1;
  const left = mergeSort(arr.slice(0, mid));
  const right = mergeSort(arr.slice(mid));
  return merge(left, right);
};

const merge = (a, b) => {
  const out = [];
  while (a.length && b.length) {
    out.push(a[0] <= b[0] ? a.shift() : b.shift());
  }
  return [...out, ...a, ...b];
};

const main = () => {
  const N = 200000; // use smaller N, merge-sort is heavy
  const nums = Array.from({ length: N }, () => Math.random());
  const t0 = performance.now();
  mergeSort(nums);
  const t1 = performance.now();
  console.log(`O(n log n) demo with N=${N} took ${(t1 - t0).toFixed(3)} ms`);
};

main();

File: golang/4-onlogn/4-OnLogN_merge_sort.go
/*
   OnLogN_merge_sort.go
   Big-O: O(n log n)
   Classic merge-sort: log n levels of splitting × n work each level.
*/

package on2

import (
	"fmt"
	"math/rand"
	"time"
)

func mergeSort(arr []float64) []float64 {
	if len(arr) <= 1 {
		return arr
	}

	mid := len(arr) >> 1
	left := mergeSort(arr[:mid])
	right := mergeSort(arr[mid:])

	return merge(left, right)
}

func merge(a, b []float64) []float64 {
	result := make([]float64, 0, len(a)+len(b))
	i, j := 0, 0

	for i < len(a) && j < len(b) {
		if a[i] <= b[j] {
			result = append(result, a[i])
			i++
		} else {
			result = append(result, b[j])
			j++
		}
	}

	// Append remaining elements
	result = append(result, a[i:]...)
	result = append(result, b[j:]...)

	return result
}

func Demo() {
	N := 200000 // use smaller N, merge-sort is heavy
	nums := make([]float64, N)
	for i := 0; i < N; i++ {
		nums[i] = rand.Float64()
	}

	t0 := time.Now()
	mergeSort(nums)
	t1 := time.Now()

	fmt.Printf("O(n log n) demo with N=%d took %.3f ms\n", N, float64(t1.Sub(t0).Microseconds())/1000)
}
